### PR TITLE
Fixed #27111 -- Prevented KeyError when USERNAME_FIELD is not in UserCreationForm.fields

### PR DESCRIPTION
**Pull request by berkerpeksag**
https://code.djangoproject.com/ticket/27111

### CODE REVIEW

File: django/contrib/auth/forms.py [hunk 1]
```diff
@@ -94,7 +94,8 @@ class Meta:
 
     def __init__(self, *args, **kwargs):
         super(UserCreationForm, self).__init__(*args, **kwargs)
-        self.fields[self._meta.model.USERNAME_FIELD].widget.attrs.update({'autofocus': ''})
+        if hasattr(self, 'fields') and self._meta.model.USERNAME_FIELD in self.fields:
```
Comment (jdufresne) [inline django/contrib/auth/forms.py]: I think the `hasattr` is unnecessary as the check is after `super().__init__()`. A child class of `BaseForm` will always have a `fields` instance variable, right? Removing the `hasattr` does not cause any tests to fail.

Comment (berkerpeksag) [inline django/contrib/auth/forms.py]: Good point, removed it. Thanks!

File: tests/auth_tests/test_forms.py [hunk 2]
```diff
@@ -209,6 +212,28 @@ class Meta(UserCreationForm.Meta):
...
         self.assertTrue(form.is_valid())
 
+    def test_custom_form_hidden_username_field(self):
+        class CustomUserCreationForm(UserCreationForm):
+            class Meta(UserCreationForm.Meta):
+                model = CustomUserWithUUIDAsUsername
+                # We deliberately left USERNAME_FIELD outside
+                fields = ('email',)
+
+            def save(self, commit=True):
```
Comment (timgraham) [inline tests/auth_tests/test_forms.py]: Needed? Couldn't the test use an existing model and omit the USERNAME_FIELD?

**Comment by timgraham** [comment]
Please add a release note for 1.10.1:
```
* Reallowed subclassing ``UserCreationForm`` without ``USERNAME_FIELD`` in
  ``Meta.fields`` (:ticket:`27111`).
```

File: tests/auth_tests/test_forms.py [hunk 3]
```diff
@@ -209,6 +212,28 @@ class Meta(UserCreationForm.Meta):
         form = CustomUserCreationForm(data)
         self.assertTrue(form.is_valid())
 
+    def test_custom_form_hidden_username_field(self):
+        class CustomUserCreationForm(UserCreationForm):
+            class Meta(UserCreationForm.Meta):
+                model = CustomUserWithUUIDAsUsername
+                # We deliberately left USERNAME_FIELD outside
+                fields = ('email',)
```
Comment (timgraham) [inline tests/auth_tests/test_forms.py]: I'd say: `fields = ('email',)  # without USERNAME_FIELD`

File: tests/auth_tests/test_forms.py (hunk 2 above)
Comment (berkerpeksag) [inline tests/auth_tests/test_forms.py]: Yes, it's definitely not needed. Updated the PR to re-use `CustomUserWithoutIsActiveField` instead.

**Comment by berkerpeksag** [comment]
I've added a release note, thanks!

**Pull request merged**